### PR TITLE
ci: Add automated benchmarking workflow

### PR DESCRIPTION
**Pull request by daniel-mills-cqc**
# Description

Adds a workflow to run benchmarks on each new PR. The version of pytket built from the PR is used in the benchmarks. At the moment this is a simple comparison to Qiskit on quantum volume circuits.

I would like to have a wider conversation about the compilers to compare agains, and the circuits to compile. The outcome of those conversation would feature in later PRs.

I also note that the benchmarks are quite slow. I will experiment with parallelisation in future PRs.

# Related issues

N/A

# Checklist

- [x] I have performed a self-review of my code.
- [x] I have commented hard-to-understand parts of my code.
- [x] I have made corresponding changes to the public API documentation.
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have updated the changelog with any user-facing changes.

### CODE REVIEW

File: .github/workflows/pytket_benchmarking.yml [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+
+      - name: Install dependencies
+        run: |
+          pip install pytket-*.whl
+          pip install --pre --index-url https://github_actions:${{ secrets.PRIVATE_PYPI_PASS }}@cqcpythonrepository.azurewebsites.net/simple/ pytket_benchmarking
+
+      - name: Checkout tket
+        uses: actions/checkout@v4
+        with:
+          path: tket
```
Comment (cqc-alec) [inline .github/workflows/pytket_benchmarking.yml:84]: Why is this necessary, as we are just using the downloaded wheel?

File: .github/workflows/pytket_benchmarking.yml [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    - name: Build tket C++
+      run: conan create tket --user tket --channel stable --build=missing -o boost/*:header_only=True -o tklog/*:shared=True -o tket/*:shared=True -tf ""
+
+    - name: Build wheel
+      run: |
+        conan create recipes/pybind11
+        conan create recipes/pybind11_json/all --version 0.2.13
+        cd pytket
+        # Ensure wheels are compatible with MacOS 12.0 and later:
+        export WHEEL_PLAT_NAME=macosx_12_0_arm64
```
Comment (cqc-alec) [inline .github/workflows/pytket_benchmarking.yml:48]: This shouldn't be necessary as we're testing on `macos-14`.

**Review by cqc-alec — COMMENTED** [review]
Looks good, just a couple of comments.

File: .github/workflows/pytket_benchmarking.yml (hunk 1 above)
Comment (daniel-mills-cqc) [reply .github/workflows/pytket_benchmarking.yml]: Ah yes, a bit of a left over from testing. Cheers!

File: .github/workflows/pytket_benchmarking.yml (hunk 2 above)
Comment (daniel-mills-cqc) [reply .github/workflows/pytket_benchmarking.yml:48]: Would you mind saying a little bit more about that. I'm afraid I'm not so familiar with the steps required to build the wheels.
I tried removing these two lines but see that the build then fails. I'm not sure if I should have expected that?

Comment (cqc-alec) [reply .github/workflows/pytket_benchmarking.yml:48]: Hmm, that's a surprise. Ah well, we can keep the lines there, it's not an issue.

**Review by cqc-alec — APPROVED** [review]

**Review by ferbetanzo — APPROVED** [review]
LGTM!
